google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/StreamWriterV2.java

```diff
@@ -322,7 +323,11 @@ private void appendLoop() {
 
       // TODO: Add reconnection here.
       while (!localQueue.isEmpty()) {
-        this.streamConnection.send(localQueue.pollFirst().message);
+        AppendRowsRequest preparedRequest =
+            prepareRequestBasedOnPosition(
+                localQueue.pollFirst().message, isFirstRequestInConnection);
+        this.streamConnection.send(preparedRequest);
+        isFirstRequestInConnection = false;
```

Comment:
Could there be a race on the isFirstRequestInConnection?
Also should the fields be cleared when after the first success response?

Reply:
There can NOT be a race between it is a local variable accessed in a single thread. It is not shared at all.

I thought about this. If the first request failed, the entire connection will be aborted, so we don't need to wait for the first success response. In this way, we can keep it as a local variable to avoid any concurrency issue.

Reply:
I am fine with the change now, but the first failure could be a in-stream failure (where the stream is not broken, for example https://screenshot.googleplex.com/5b4zkgK89qudYDk), maybe add a comment about the potential issue here, or open a bug for me to change every failure in the first request to be a connection failure.

Reply:
Discussed offline. This change may introduce some risk. So I'll spent a bit more time on this.

Reply:
Discussed offline, I think if we have client side schema compatibility check built in, the risk would be minimum (or be able to detect while client is writing their code, instead of runtime).

Reply:
There are a lot of concerns from different people on how the schema should be passed. So I'll park this thread for now.

Reply:
Discussed offline, we will proceed with this as step 1.

In step 2, we will accept the schema when creating StreamWriterV2 and provide a new append method.